libtailscale/vpnfacade.go

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package libtailscale
+
+import (
+	"sync"
+
+	"tailscale.com/net/dns"
+	"tailscale.com/wgengine/router"
+)
+
+// VPNFacade is an implementation of both wgengine.Router and
+// dns.OSConfigurator. When ReconfigureVPN is called by the backend, SetBoth
+// gets called.
+type VPNFacade struct {
+	SetBoth func(rcfg *router.Config, dcfg *dns.OSConfig) error
+
+	// GetBaseConfigFunc optionally specifies a function to return the current DNS
+	// config in response to GetBaseConfig.
+	//
+	// If nil, reading the current config isn't supported and GetBaseConfig()
+	// will return ErrGetBaseConfigNotSupported.
+	GetBaseConfigFunc func() (dns.OSConfig, error)
+
+	// InitialMTU is the MTU the tun should be initialized with.
+	// Zero means don't change the MTU from the default. This MTU
+	// is applied only once, shortly after the TUN is created, and
+	// ignored thereaftef.
+	InitialMTU uint32
+
+	mu        sync.Mutex     // protects all the following
+	didSetMTU bool           // if we set the MTU already
+	rcfg      *router.Config // last applied router config
+	dcfg      *dns.OSConfig  // last applied DNS config
+}
+
+// Up implements router.
```

Comment:
wgengine.Router

I'd also add some static type assertions at the package level:

```go
var (
    _ wgengine.Router = (*VPNFaçade)(nil)
    _ foo.Bar = (*VPNFaçade)(nil)
)
```